keps/sig-auth/20200507-external-tls-certificate-authenticator/README.md

```diff
@@ -0,0 +1,749 @@
+
+How will security be reviewed and by whom?
+
+How will UX be reviewed and by whom?
+
+Consider including folks that also work outside the SIG or subproject.
+-->
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your change
+are understandable.  This may include API specs (though not always required) or
+even code snippets.  If there's any ambiguity about HOW your proposal will be
+implemented, this is the place to discuss them.
+-->
+
+The complete solution for external TLS certificate signing includes two
+components:
+
+- an authentication provider within kubectl/client-go, and
+- an external plugin, which provides the actual certificate and signing
+  services.
+
+![UML sequence diagram](uml_sequence_diagram.png)
+
+The main role of the authentication provider is to pass along requests for
+obtaining a TLS certificate and signing to the external plugin, retrieve the
+responses and return them to the initial caller.
+
+Users would be required to install an external plugin (possibly along with
+dependencies) on their workstation.
+
+### Configuration
+
+Configuration of the both components (authentication provider and external
+plugin) is specified in the kubeconfig files as part of the user fields. The
+authentication provider is responsible for reading the configuration from the
+kubeconfig file and exposing the configuration parameters to the external
+plugin.
+
+The internal authentication provider requires only a single parameter, the path
+to the external plugin `pathExec`.
+
+<!-- TODO: Relative command paths are interpreted as relative to the directory of the config file. If KUBECONFIG is set to /home/jane/kubeconfig and the exec command is ./bin/example-client-go-exec-plugin, the binary /home/jane/bin/example-client-go-exec-plugin is executed. -->
+
+The remaining of the parameters are specific for the external plugin and
+protocol in use. In order to increase flexibility of the solution and support
+for multiple protocols, the authentication provider will pass all the
+configuration parameters specified in the kubeconfig file to the external plugin
+in a form of string-to-string mapping (key-value pairs).
+
+In case of the PKCS#11 protocol, the following parameters are in use:
+
+- `pathLib` - a path to the library used by the authentication protocol
+  (mandatory),
+- `slotId` - an identifier of the slot (mandatory),
+- `objectId` - an identifier of the object (mandatory),
+- `pin` - a PIN code used when accessing the private key during the signing
+  operation (optional, if not provided, the user will be asked to provide the
+  PIN during each signing operation).
+
+An excerpt from an exemplary kubeconfig file:
+
+```yaml
+apiVersion: v1
+kind: Config
+users:
+- name: my-user
+  user:
+    auth-provider:
+      name: externalSigner
+      config:
+        pathExec: /path/to/externalSigner  
+        pathLib: /path/to/library.so        # library used by the external plugin
```

Comment:
will /path/to normally require root?
if not, what prevents a malicious actor on the same host from overloading the signer to MITM the data exchange?